### PR TITLE
z_kankyo pt 1

### PR DESCRIPTION
**Pull request by hensldm**
Starts to bring over z_kankyo over from Decompollaborate (https://github.com/Decompollaborate/mm/tree/z_kankyo_new). This involves mainly simple functions, along with bringing over the new header, and importing data and bss.

Thanks to everyone who has helped worked on the Kankyo branch :)

### CODE REVIEW

File: include/segment_symbols.h [hunk 1]
```diff
@@ -32,6 +32,11 @@
 #define SEGMENT_BSS_END(segment) (_ ## segment ## SegmentBssEnd)
 #define SEGMENT_BSS_SIZE(segment) ((uintptr_t)SEGMENT_BSS_END(segment) - (uintptr_t)SEGMENT_BSS_START(segment))
 
+#define ROM_FILE(name) \
+    { (uintptr_t)_##name##SegmentRomStart, (uintptr_t)_##name##SegmentRomEnd }
```
Comment (AngheloAlf) [inline include/segment_symbols.h:36]: What about this instead?
```suggestion
    { (uintptr_t) SEGMENT_ROM_START(name), (uintptr_t)SEGMENT_ROM_END(name) }
```

File: src/code/z_kankyo.c [hunk 2]
```diff
@@ -1,163 +1,992 @@
...
+    Environment_LerpRGB8(play->envCtx.lightSettings.fogColor, to, lerp, play->envCtx.adjLightSettings.fogColor);
+}
+
+void Environment_LerpFog(PlayState* play, s16 fogNearTarget, s16 fogFarTarget, f32 lerp) {
+    play->envCtx.adjLightSettings.fogNear = (fogNearTarget - play->envCtx.lightSettings.fogNear) * lerp;
+    play->envCtx.adjLightSettings.zFar = (fogFarTarget - play->envCtx.lightSettings.zFar) * lerp;
+}
+
+// Repurposed from OoT to be more general
+u32 Environment_GetBgsDayCount(void) {
```
Comment (AngheloAlf) [inline src/code/z_kankyo.c:842]: What does "bgs" mean?

Comment (hensldm) [reply src/code/z_kankyo.c:842]: Its from OoT. Based on usage I think it is `Biggoron Sword` but I think Engineer would know more.

Comment (fig02) [reply src/code/z_kankyo.c:842]: yes it is biggoron sword. in oot it keeps track of the days that have passed since you gave claim check, cause it takes 3 days for him to make the sword

if it gets used for other things in this game, a more general name could be thought of i suppose

Comment (engineer124) [reply src/code/z_kankyo.c:842]: Yes exactly, it was just named after OoT early in WIP to help with cross-referencing the files since the function is identical, but in MM is repurposed to be more general. It was just never named to be more general but probably should be

Comment (hensldm) [reply src/code/z_kankyo.c:842]: Posted in #mm-names-discussion

File: src/code/z_demo.c [hunk 3]
```diff
@@ -13,6 +11,28 @@ static s16 sBssPad;
...
+struct Dummy207 { int x; };
+struct Dummy208 { int x; };
+struct Dummy209 { int x; };
+struct Dummy210 { int x; };
+struct Dummy211 { int x; };
+struct Dummy212 { int x; };
+struct Dummy213 { int x; };
+struct Dummy214 { int x; };
+struct Dummy215 { int x; };
+// clang-format on
```
Comment (hensldm) [inline src/code/z_demo.c:34]: Did want to bring attention to this. The only way I could get bss to match.

Comment (AngheloAlf) [reply src/code/z_demo.c:34]: You had to put those _after_ the bss variables? It doesn't work before them?

Comment (hensldm) [reply src/code/z_demo.c:34]: Its not after `all` of them. These are to push `sSeqId` to the top.

**Review by AngheloAlf — APPROVED** [review]

File: src/code/z_play.c [hunk 4]
```diff
@@ -21,7 +22,6 @@ u8 sMotionBlurStatus;
 #include "buffers.h"
 #include "idle.h"
 #include "sys_cfb.h"
-#include "z64bombers_notebook.h"
```
Comment (engineer124) [inline src/code/z_play.c:24]: why is this removed from z_play?

Comment (hensldm) [reply src/code/z_play.c]: Not sure, probably just an accident. Will add it back.

**Review by engineer124 — APPROVED** [review]

File: src/code/z_kankyo.c [hunk 5]
```diff
@@ -1,163 +1,992 @@
...
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_kankyo/func_800FD78C.s")
+u32 Environment_GetStormState(PlayState* play) {
+    u32 stormState = play->envCtx.stormState;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_kankyo/func_800FD858.s")
+    if ((play->sceneId == SCENE_OMOYA) && (play->roomCtx.curRoom.num == 0)) {
+        stormState = ((gSaveContext.save.day >= 2) && !CHECK_WEEKEVENTREG(WEEKEVENTREG_DEFENDED_AGAINST_THEM))
+                         ? STORM_STATE_ON
+                         : STORM_STATE_OFF;
+    }
```
Comment (isghj5) [inline src/code/z_kankyo.c:903]: Wait some weather is determined by aliens?

File: src/code/z_kankyo.c [hunk 6]
```diff
@@ -1,163 +1,992 @@
...
+        ret = true;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_kankyo/func_800FE4B8.s")
+    return ret;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_kankyo/func_800FE590.s")
+u32 func_800FE610(PlayState* play) {
+    return 0;
```
Comment (isghj5) [inline src/code/z_kankyo.c:952]: :thinking:

File: src/code/z_lights.c [hunk 7]
```diff
@@ -257,15 +257,15 @@ void LightContext_Init(PlayState* play, LightContext* lightCtx) {
...
+    lightCtx->ambientColor[2] = b;
 }
 
 void LightContext_SetFog(LightContext* lightCtx, u8 r, u8 g, u8 b, s16 near, s16 far) {
-    lightCtx->fogColor.r = r;
-    lightCtx->fogColor.g = g;
-    lightCtx->fogColor.b = b;
+    lightCtx->fogColor[0] = r;
+    lightCtx->fogColor[1] = g;
+    lightCtx->fogColor[2] = b;
```
Comment (isghj5) [inline src/code/z_lights.c:268]: The colors have reverted to array from rgba?

File: src/overlays/actors/ovl_Demo_Kankyo/z_demo_kankyo.c [hunk 8]
```diff
@@ -56,21 +56,21 @@ void DemoKakyo_LostWoodsSparkleActionFunc(DemoKankyo* this, PlayState* play) {
...
+        play->envCtx.precipitation[PRECIP_SNOW_MAX] += 16;
     }
 
     // note: DemoKankyo can crash if placed in an area that snows (ObjectKankyo)
     // because they both use unk_F2 as an effect counter,
     // causing DemoKankyo to write beyond its efffect array boundry
     // this crash can occur if the two actors are in different scenes connected by an exit
     // e.g. if you add DemoKankyo to GoronShrine, you will crash entering/leaving through door
-    for (i = 0; i < play->envCtx.unk_F2[3]; i++) {
+    for (i = 0; i < play->envCtx.precipitation[PRECIP_SNOW_MAX]; i++) {
```
Comment (isghj5) [inline src/overlays/actors/ovl_Demo_Kankyo/z_demo_kankyo.c:73]: This comment could probably be updated with the change to `.precipitation[PRECIP_SNOW_MAX]`

File: src/overlays/actors/ovl_En_Weather_Tag/z_en_weather_tag.c [hunk 9]
```diff
@@ -243,7 +237,7 @@ void EnWeatherTag_Die(EnWeatherTag* this, PlayState* play) {
 //  poisoned swamp: placed behind the water fall from ikana
 // this tag stops spawning after STT cleared?
 void func_80966B08(EnWeatherTag* this, PlayState* play) {
-    if (func_80966608(this, play, 0, 0, play->envCtx.lightConfig, 5, 100, 2) || (gWeatherMode == 2)) {
+    if (func_80966608(this, play, 0, 0, play->envCtx.lightConfig, 5, 100, 2) || (gWeatherMode == WEATHER_MODE_2)) {
```
Comment (isghj5) [inline src/overlays/actors/ovl_En_Weather_Tag/z_en_weather_tag.c:240]: Bah, the mystery continues.

File: include/z64environment.h [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+
+typedef struct {
+    /* 0x00 */ u8 ambientColor[3];
+    /* 0x03 */ s8 light1Dir[3];
+    /* 0x06 */ u8 light1Color[3];
+    /* 0x09 */ s8 light2Dir[3];
+    /* 0x0C */ u8 light2Color[3];
+    /* 0x0F */ u8 fogColor[3];
+    /* 0x12 */ s16 fogNear;
+    /* 0x14 */ s16 zFar;
```
[the remark this answers is not in the document]
Comment (fig02) [reply include/z64environment.h:166]: comments from oot would be good to port

File: src/code/z_lights.c (hunk 7 above)
Comment (fig02) [reply src/code/z_lights.c:268]: the environment dev loves using arrays for colors, and it was needed to match in some cases (in oot and probably here)

File: include/z64environment.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (isghj5) [reply include/z64environment.h:166]: In hindsight, this is just PR 1. Docs can wait.

File: src/code/z_kankyo.c (hunk 5 above)
Comment (hensldm) [reply src/code/z_kankyo.c:903]: Yeah, the barn. If you lose than the roof is gone so would need to have the same storm state as outside.

File: include/z64environment.h [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+    /* 0x4 */ SANDSTORM_DISSIPATE,
+    /* 0x5 */ SANDSTORM_5,
+    /* 0x6 */ SANDSTORM_6,
+    /* 0x7 */ SANDSTORM_7,
+    /* 0x8 */ SANDSTORM_8,
+    /* 0x9 */ SANDSTORM_9,
+    /* 0xA */ SANDSTORM_A,
+    /* 0xB */ SANDSTORM_B,
+    /* 0xC */ SANDSTORM_C,
+    /* 0xD */ SANDSTORM_D
```
[the remark this answers is not in the document]
Comment (hensldm) [reply include/z64environment.h:134]: Most of these were actually added in MM (only the named ones actually come from OoT).

File: include/z64environment.h [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+} SandstormState;
+
+typedef struct LightningStrike {
+    /* 0x00 */ u8 state;
+    /* 0x01 */ u8 flashRed;
+    /* 0x02 */ u8 flashGreen;
+    /* 0x03 */ u8 flashBlue;
+    /* 0x04 */ u8 flashAlphaTarget;
```
[the remark this answers is not in the document]
Comment (hensldm) [reply include/z64environment.h:142]: Discussed in the z_kankyo thread on discord

File: include/z64environment.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (hensldm) [reply include/z64environment.h:166]: Pretty simple to port the comments over so did.

File: src/code/z_lights.c (hunk 7 above)
Comment (hensldm) [reply src/code/z_lights.c:268]: Can confirm. At least one function iterates over them in a for loop. Just not in this PR.

File: src/code/z_kankyo.c (hunk 5 above)
Comment (isghj5) [reply src/code/z_kankyo.c:903]: Oh that makes sense, I was assuming the storm was caused by the aliens or something but yeah, the roof being gone.

**Review by isghj5 — APPROVED** [review]
